cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -796,10 +799,52 @@ private Node getAPriorDefinition(SsaDefOrUse defOrUse) {
   )
 }
 
+private predicate inOut(FIO::FunctionInput input, FIO::FunctionOutput output) {
+  exists(int indirectionIndex |
+    input.isQualifierObject(indirectionIndex) and
+    output.isQualifierObject(indirectionIndex)
+    or
+    exists(int i |
+      input.isParameterDeref(i, indirectionIndex) and
+      output.isParameterDeref(i, indirectionIndex)
+    )
+  )
+}
+
+/**
+ * Holds if there should not be use-use flow out of `n` (or a conversion that
+ * flows to `n`).
+ */
+private predicate modeledFlowBarrier(Node n) {
+  exists(
+    FIO::FunctionInput input, FIO::FunctionOutput output, CallInstruction call,
+    PartialFlow::PartialFlowFunction partialFlowFunc
+  |
+    n = callInput(call, input) and
+    inOut(input, output) and
+    exists(callOutput(call, output)) and
+    partialFlowFunc = call.getStaticCallTarget() and
+    not partialFlowFunc.isPartialWrite(output)
+  |
+    call.getStaticCallTarget().(DataFlow::DataFlowFunction).hasDataFlow(_, output)
+    or
+    call.getStaticCallTarget().(Taint::TaintFunction).hasTaintFlow(_, output)
+  )
+  or
+  exists(Operand operand, Instruction instr, Node n0, int indirectionIndex |
+    modeledFlowBarrier(n0) and
+    nodeHasInstruction(n0, instr, indirectionIndex) and
+    conversionFlow(operand, instr, false, _) and
+    nodeHasOperand(n, operand, indirectionIndex)
+  )
+}
+
 /** Holds if there is def-use or use-use flow from `nodeFrom` to `nodeTo`. */
 predicate ssaFlow(Node nodeFrom, Node nodeTo) {
   exists(Node nFrom, boolean uncertain, SsaDefOrUse defOrUse |
-    ssaFlowImpl(defOrUse, nFrom, nodeTo, uncertain) and nodeFrom != nodeTo
+    ssaFlowImpl(defOrUse, nFrom, nodeTo, uncertain) and
+    not modeledFlowBarrier(nFrom) and
```

Comment:
Let's assume you have this snippet:
```cpp
void test() {
  char* p = taint(); // (1)
  foo(p); // (2)
  sink(p); // (3)
}
```
Let's consider three scenarios:

1. You have no model for `foo`
2. You have a dataflow model for `foo` that looks like:
```ql
override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
  input.isParameterDeref(0) and
  output.isParameterDeref(0)
}
```
3. You have a dataflow model for `foo` that looks like the above, but also has:
```ql
override predicate isPartialWrite(FunctionOutput output) { output.isParameterDeref(0) }
```

The behavior for each case will be:

## You have no model for `foo`
There will be flow from `taint()` to `p` (in `// (2)`), and from `p` (in `// (2)`) to `p` (in `// (3)`). See here:
<img width="758" alt="Screenshot 2024-02-22 at 14 47 50" src="https://github.com/github/codeql/assets/4527323/e5d11b6d-66f3-40ae-b9e0-caac898b22b3">

## You have a dataflow model for `foo`, but no `isPartialWrite` override

There will be flow from `taint()` to `p` (in `// (2)`), and from `p` (in `// (2)`) _to the output argument of `foo`_, and from the output argument of `foo` to `p` (in `// (3)`. See here:
<img width="696" alt="Screenshot 2024-02-22 at 14 50 17" src="https://github.com/github/codeql/assets/4527323/e7467c64-f0ae-4ea6-a73a-aca04389a55a">

That is, now that the model for `foo` implicitly blocks flow it means that flow will have to go through `foo` in order to reach `sink`.

## You have a dataflow model for `foo`, and you also override `isPartialWrite`.

Since the flow is now specified as a partial write it means you get both of the flows from before: 
<img width="990" alt="Screenshot 2024-02-22 at 14 51 48" src="https://github.com/github/codeql/assets/4527323/02ca9762-5cb5-4d3b-9ae0-baf0bed05758">

Reply:
I should add: The paths above have some duplication since I disable any kind of path compression in my example.